nemo/collections/asr/parts/submodules/fast_rnnt_greedy_decoding.py

```diff
@@ -0,0 +1,440 @@
+from nemo.collections.common.parts.rnn import label_collate
+from nemo.core.classes import Typing, typecheck
+from nemo.core.neural_types import AcousticEncodedRepresentation, ElementType, HypothesisType, LengthsType, NeuralType
+from nemo.utils import logging
+
+
+def ASSERT_DRV(err):
+    if isinstance(err, cuda.CUresult):
+        if err != cuda.CUresult.CUDA_SUCCESS:
+            raise RuntimeError("Cuda Error: {}".format(err))
+    elif isinstance(err, nvrtc.nvrtcResult):
+        if err != nvrtc.nvrtcResult.NVRTC_SUCCESS:
+            raise RuntimeError("Nvrtc Error: {}".format(err))
+    else:
+        raise RuntimeError("Unknown error type: {}".format(err))
+
+
+def cu_call(f_call_out):
+    error, *others = f_call_out
+    if error != cudart.cudaError_t.cudaSuccess:
+        # import ipdb; ipdb.set_trace()
+        raise Exception(f"CUDA failure! {error}")
+    else:
+        # print("GALVEZ:", others)
+        return tuple(others)
+
+
+def run_nvrtc(kernel_string, kernel_name):
+    err, prog = nvrtc.nvrtcCreateProgram(str.encode(kernel_string), b"while_loop_conditional.cu", 0, [], [])
+
+    ASSERT_DRV(err)
+
+    # Compile program
+    # Not specifying --gpu-architecture will default us to a fairly low compute capability, which is a safe bet.
+    # Otherwise, there are ways to query the current device's compute capability.
+    # https://stackoverflow.com/questions/48283009/nvcc-get-device-compute-capability-in-runtime
+    opts = []
+    (err,) = nvrtc.nvrtcCompileProgram(prog, len(opts), opts)
+    err, size = nvrtc.nvrtcGetProgramLogSize(prog)
+    buf = b" " * size
+    (err,) = nvrtc.nvrtcGetProgramLog(prog, buf)
+    # print(buf.decode("utf-8"))
+    ASSERT_DRV(err)
+
+    # Get PTX from compilation
+    err, ptxSize = nvrtc.nvrtcGetPTXSize(prog)
+    ASSERT_DRV(err)
+    ptx = b" " * ptxSize
+    (err,) = nvrtc.nvrtcGetPTX(prog, ptx)
+    ASSERT_DRV(err)
+
+    # print("GALVEZ:PTX=")
+    # print(ptx.decode("utf-8"))
+    ptx = np.char.array(ptx)
+    err, module = cuda.cuModuleLoadData(ptx.ctypes.data)
+    ASSERT_DRV(err)
+    err, kernel = cuda.cuModuleGetFunction(module, kernel_name)
+    ASSERT_DRV(err)
+
+    return kernel
+
+
+def create_outer_for_loop_kernel():
+    kernel_string = r"""\
+    typedef __device_builtin__ unsigned long long cudaGraphConditionalHandle;
+
+    extern "C" __device__ __cudart_builtin__ void cudaGraphSetConditional(cudaGraphConditionalHandle handle, unsigned int value);
+
+    extern "C" __global__
+    void for_loop_conditional(cudaGraphConditionalHandle handle, const long *time_idx, const long *trip_count)
+    {
+     cudaGraphSetConditional(handle, *time_idx < *trip_count);
+    }
+    """
+    return run_nvrtc(kernel_string, b"for_loop_conditional")
+
+
+# Observations: If cudaGraphSetConditional is true once, the kernel never completes...
+# The GPU is doing *something*. I'm just not sure what...
+# No way to query that at runtime...
```

Comment:
Uhh is this still an issue or was it resolved during development?

Reply:
resolved a long time ago. Sorry, I write comments to myself during development.